### PR TITLE
Feature: Edge acyclicity speedup

### PR DESCRIPTION
**Pull request by tadhgokeeffe**
## Description and Motivation
Added a check to ensure that nodes aren't checked more than once when checking for acyclicity.

## Additional Context
Marcin identified issue with adding edges becoming extremely slow in highly connected graphs

## How Has This Been Tested?
Tested in notebook
Screenshot 2023-10-31 at 14 02 39
Added unittest to make sure runs in a reasonable amount of time

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [x] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [ ] I have implemented all requirements? (see JIRA, project documentation)
- [ ] I am affecting someone else's work? If so: I have added those people as reviewers?
- [x] I have added comments to all the bits that are hard to follow
- [x] At a bare minimum, I tested this manually
- [x] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [x] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [x] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: cai_causal_graph/causal_graph.py
Comment (andrew-causalens) [inline cai_causal_graph/causal_graph.py:1]: Can you add another unit test that runs it on larger graph and works? Just so we can see in future if we accidentally slow it down again. Will see that tests take forever to run.

**Review by matteo-causalens — APPROVED** [review]
LGTM

File: cai_causal_graph/causal_graph.py
Comment (tadhgokeeffe) [reply cai_causal_graph/causal_graph.py:1]: Will do. Ill add Marcins example

**Review by steven-causalens — APPROVED** [review]
LGTM, thanks @tadhgokeeffe

File: changelog.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## NEXT
+
+- Fixed a bug to improve checking for cycles when adding edges by avoiding repeated checks.
```
Comment (maxelliott-causalens) [inline changelog.md:5]: "Fixed a bug" implies that the behaviour was wrong, people might think we were allowing cycles before this in some cases. Would just say "Improved performance of checking for cycles..."

**Review by maxelliott-causalens — APPROVED** [review]
Looks good, would suggest changing the changelog slightly though for clarity

**Review by mmiklitz — APPROVED** [review]

**Pull request merged**